runtime/common/src/main/kotlin/kotlinx/serialization/context/Context.kt

```diff
@@ -3,45 +3,175 @@
+     *
+     * Given [concreteSerializer] will be used in polymorphic serialization
+     * if statically known type of property is [basePolyType] and runtime type is a [concreteClass].
+     *
+     * For serialization process, classname is determined using name of a serializer descriptor.
+     *
+     * All standard types and collections are registered in advance as a subtypes in a scope of [kotlin.Any].
+     */
+    fun <Base : Any, Sub : Base> registerPolymorphicSerializer(
+        basePolyType: KClass<Base>,
+        concreteClass: KClass<Sub>,
+        concreteSerializer: KSerializer<Sub>
+    )
 }
 
 class MutableSerialContextImpl(private val parentContext: SerialContext? = null): MutableSerialContext {
 
-    private val classMap: MutableMap<KClass<*>, KSerializer<*>> = hashMapOf()
+    private val classMap: SerializersMap = hashMapOf()
+
+    private val polyMap: MutableMap<KClass<*>, SerializersMap> = hashMapOf()
+    private val inverseClassNameMap: MutableMap<KClass<*>, MutableMap<String, KSerializer<*>>> = hashMapOf()
 
     override fun <T: Any> registerSerializer(forClass: KClass<T>, serializer: KSerializer<T>) {
         classMap[forClass] = serializer
     }
 
+    override fun <Base : Any, Sub : Base> registerPolymorphicSerializer(
+        basePolyType: KClass<Base>,
+        concreteClass: KClass<Sub>,
+        concreteSerializer: KSerializer<Sub>
+    ) {
+        val name = concreteSerializer.descriptor.name
+        polyMap.getOrPut(basePolyType, ::hashMapOf)[concreteClass] = concreteSerializer
+        inverseClassNameMap.getOrPut(basePolyType, ::hashMapOf)[name] = concreteSerializer
+    }
+
+    override fun <T : Any> resolveFromBase(basePolyType: KClass<T>, obj: T): KSerializer<out T>? {
+        if (!isInstance(basePolyType, obj)) return null
+        (if (basePolyType == Any::class) StandardSubtypesOfAny.getSubclassSerializer(obj) else null)?.let { return it as KSerializer<out T> }
+        return polyMap[basePolyType]?.get(obj::class) as? KSerializer<out T>
+    }
+
+    override fun <T : Any> resolveFromBase(basePolyType: KClass<T>, serializedClassName: String): KSerializer<out T>? {
+        (if (basePolyType == Any::class) StandardSubtypesOfAny.getDefaultDeserializer(serializedClassName) else null)?.let { return it as KSerializer<out T> }
+        return inverseClassNameMap[basePolyType]?.get(serializedClassName) as? KSerializer<out T>
+    }
+
     override fun <T : Any> getByValue(value: T): KSerializer<T>? {
-//        if (value == null) throw SerializationException("Cannot determine class for value $value")
-        val t: T = value
-        val klass = t::class
+        val klass = value::class
         return get(klass) as? KSerializer<T>
     }
 
     override fun <T: Any> get(kclass: KClass<T>): KSerializer<T>? = classMap[kclass] as? KSerializer<T> ?: parentContext?.get(kclass)
 }
 
+
+internal object StandardSubtypesOfAny {
```

Comment:
A lot of classes for a single file, please split.
Contexts have a separate package, so it shouldn't be a big deal, but users will see only "public" documented API without undocumented implementation.

Reply:
Currently, this is a common approach for `kotlinx.serialization`, but one public entity <-> one file is much easier to understand and separate implementation from declarations